<launch>
  <!-- for viso -->
  <arg name="stereo" default="multisense" />
  <arg name="image" default="image_rect" />
  <arg name="odom_frame_id" default="viso_odom" />
  <arg name="base_link_frame_id" default="BODY" />
  <arg name="sensor_frame_id" default="left_camera_optical_frame" />
  <arg name="output_odom_topic" default="viso_odom" />
  <arg name="publish_viso_tf" default="false" />
  <arg name="invert_viso_tf" default="false" />
  <!-- for ekf -->
  <arg name="use_robot_pose_ekf" default="true" />
  <arg name="odom_used" default = "true" />
  <arg name="odom_data" default="/odom" />
  <arg name="imu_used" default = "false" />
  <arg name="imu_data" default="/imu" />

  <!-- viso -->
  <node pkg="viso2_ros" type="stereo_odometer" name="viso_node"
        output="screen" respawn="true">
    <remap from="stereo" to="$(arg stereo)" />
    <remap from="image" to="$(arg image)" />
    <remap from="~odometry" to="$(arg output_odom_topic)_raw" />
    <param name="~base_link_frame_id" value="$(arg base_link_frame_id)" />
    <param name="~odom_frame_id" value="$(arg odom_frame_id)_raw"/>
    <param name="~sensor_frame_id" value="$(arg sensor_frame_id)"/>    
    <param name="~publish_tf" value="$(arg publish_viso_tf)" />
    <param name="~invert_tf" value="$(arg invert_viso_tf)" />
    <param name="~ref_frame_change_method" value="1"/>
  </node>

  <node pkg="jsk_robot_startup" type="CameraToBaseOffset.py" name="viso_camera_to_base_offset"
        output="screen" respawn="true">
    <remap from="~source_odom" to="$(arg output_odom_topic)_raw" />
    <remap from="~init_signal" to="/odom_init_trigger" />
    <remap from="~output" to="$(arg output_odom_topic)" />    
    <param name="~base_link_frame" value="$(arg base_link_frame_id)" />
    <param name="~camera_frame" value="$(arg sensor_frame_id)"/>
    <param name="~odom_frame" value="$(arg odom_frame_id)"/>
    <param name="~publish_tf" value="$(arg publish_viso_tf)" />
  </node>
  
  <!-- EKF -->
  <group if="$(arg use_robot_pose_ekf)" >
    <node pkg="robot_pose_ekf" type="robot_pose_ekf" name="viso_robot_pose_ekf">
      <param name="output_frame" value="$(arg odom_frame_id)"/>
      <param name="base_footprint_frame" value="$(arg base_link_frame_id)"/>
      <param name="freq" value="20.0"/>
      <param name="sensor_timeout" value="1.0"/>
      <param name="odom_used" value="$(arg odom_used)"/>
      <param name="odom" value="$(arg odom_data)"/>
      <param name="imu_used" value="$(arg imu_used)"/>
      <remap from="imu_data" to="$(arg imu_data)" />
      <param name="vo_used" value="true"/>
      <remap from="vo" to="$(arg output_odom_topic)" />
      <param name="debug" value="false"/>
      <param name="self_diagnose" value="false"/>
      <!-- <param name="publish_tf" value="false"/> -->
    </node>
    <!-- visualization -->
    <node pkg="jsk_pcl_ros" type="pose_with_covariance_stamped_to_gaussian_pointcloud" 
          name="viso_gaussian_point_cloud" output="screen" >
      <remap from="~input" to="/viso_robot_pose_ekf/odom_combined" />
    </node>
  </group>

</launch>
